File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    private int answer = 0;
    public int solution(int k, int[][] dungeons) {
       
        boolean[] visited = new boolean[dungeons.length];
        backtrack( k, dungeons, visited, 0);    
        return answer;
    }

    private void backtrack(int k, int[][] dungeons, boolean[] visited, int cnt){
       
        for(int i=0; i<dungeons.length;i++){
            //탐험 가능한 조건
            if(dungeons[i][0]<=k && !visited[i]){ 
                visited[i]=true;
                backtrack(k-dungeons[i][1], dungeons, visited, cnt+1);
                visited[i]=false;
            }
        }
         answer = Math.max(answer, cnt);
    }   
}

File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include<bits/stdc++.h>
using namespace std;
int max_cnt=0;

void dfs(int k, vector<vector<int>> &dungeons, vector<bool> &visited, int cnt){
    max_cnt = max(cnt, max_cnt);
    
   
    for(int i=0; i<dungeons.size(); i++ ){
        if(k>=dungeons[i][0] && !visited[i]){
            visited[i]=true;
            cout<<i <<" ";
            dfs(k-dungeons[i][1], dungeons,visited,cnt+1);
            visited[i] =false;
           
        }
    }
}
int solution(int k, vector<vector<int>> dungeons) {
    int answer = -1;
    int n = dungeons.size();
    vector<bool> visited(n,false);
    dfs(k,dungeons,visited,0);
    
    answer = max_cnt;
    return answer;
}